nrcemt/qeels/engine/peak_detection.py

```diff
@@ -252,6 +252,107 @@ def draw_plasmon(spectrogram, peak_position_y, q_pixel, e_pixel):
+    images = []
+
+    # loop through different rows
+    for i in range(0, 6, 2):
+        # retrieve data for more use later on
+        x1 = plasmon_array[i][0]
+        y1 = plasmon_array[i][1]
+        x2 = plasmon_array[i+1][0]
+        y2 = plasmon_array[i+1][1]
+        width = width_array[int(i/2)]
+        detect = detect_array[int(i/2)]
+        [spectrogram_width, spectrogram_height] = np.shape(spectrogram)
+
+        # Needs better names
+        is_filled_1 = x1 > 0 or y1 > 0
+        is_filled_2 = x2 > 0 or y2 > 0
+
+        # if both values are entered
+        if detect and is_filled_1 and is_filled_2:
+            # DOUBLE CHECK, BUT I DONT THINK THE RESULT FROM THIS ARE USED
+            # calculated_corners = compute_rect_corners(x1, y1, x2, y2, width)
+
+            rotation_angle_rad, rotation_angle_degrees = calc_angle(
+                x1, y1,
+                x2, y2
+            )
+
+            # apply rotation to the points
+            x1, y1, x2, y2 = rotate_points(
+                x1, y1, x2, y2,
+                rotation_angle_rad,
+                spectrogram_width,
+                spectrogram_height
+            )
+
+            if y1 > y2:
+                temp = y1
+                y1 = y2
+                y2 = temp
+
+            # rotate image so plasmon is vertical
+            # differntiation between matlab and original
+            spectrogram_rotated = rotate_spectrogram(
+                spectrogram,
+                rotation_angle_degrees
+            )
+
+            # Find absolute value of image
+            spectrogram_signal = np.absolute(spectrogram_rotated)
+
+            peak_position_x, peak_position_y, peak_image = mark_peaks(
+                x1, y1, y2, spectrogram_signal, spectrogram, average_pixel,
+                width, rotation_angle_rad, spectrogram_height,
+                spectrogram_height
+            )
+            images.append(peak_image)
+            if i == 0:
+                e_dispersion, bulk_image = bulk_calculations(
+                    peak_position_x, peak_position_y,
+                    spectrogram
+                )
+                results_array[0] = e_dispersion
+                images.append(bulk_image)
+
+            else:
+                dispersion_q, surface_image, _ = surface_plasmon_calculations(
+                    peak_position_x, peak_position_y,
+                    e_dispersion, spectrogram
+                )
+                images.append(surface_image)
+
+                if i == 2:
+                    results_array[1] = dispersion_q
+                else:
+                    results_array[2] = dispersion_q
+
+    # need to combine the images better
+    result_image = np.zeros(spectrogram.shape)
+    results = (spectrogram)
+
+    for image in images:
+        result_image += image
+
+    for a in range(spectrogram_width):
+        for b in range(spectrogram_height):
+            if result_image[a][b] != 0:
+                results[a][b] = 5000
```

Comment:
you could feed all of the spectrograms into a np.maximum call something like `result_image = np.maximum(spectrogram, *peak_images)`. Do not use for loops like this with numpy, it is slow. If you need another way, ask.